### ISSUE TITLE
Cancel of Space migration leaves behind an empty pool

### ISSUE THREAD

**Issue by philrz** [brimdata/brimcap#85]
Repro is  in Brim commit `43b0ed5` via nightly macOS build \[https://storage.googleapis.com/brimsec-releases/brim/v0.25.0-prerelease-43b0ed51.0/macos/Brim-0.25.0-prerelease-43b0ed51.0.dmg](https://storage.googleapis.com/brimsec-releases/brim/v0.25.0-prerelease-43b0ed51.0/macos/Brim-0.25.0-prerelease-43b0ed51.0.dmg), which uses Brimcap commit 82f0140.

Before starting the first video, I've already been running GA Brim tagged `v0.24.0` and have imported data spread across five Spaces that's going to need to be migrated. In this first video I launch the new Brim at commit `43b0ed5` and let the migration start, but then click the vertical "..." and click **Cancel** when it happens to be in the middle of migrating my Space called "all.pcap". The informative pop-up accurately reflects that the migration was canceled, and clicking the "all.pcap" Space we see no data.

https://user-images.githubusercontent.com/5934157/118419523-18e2ab80-b671-11eb-8afe-acc6550774c1.mp4

The second video picks up after I've closed and reopened the app. As I've still got Spaces that have not completed migration, I once again accept the prompt to let the migration start. The pop-up gives the impression that the space "all.pcap" for which migration was cancelled mid-way on the first pass has been migrated, but at the end the pop-up says "Some Spaces not migrated". Clicking through them all, we can see that the Pool for "all.pcap" is still empty.

https://user-images.githubusercontent.com/5934157/118419623-67904580-b671-11eb-9aa1-d9412b2ab875.mp4

Finally, in the third video, I relaunch one last time and once again accept the prompt to migrate, but it just rushes to the "Some Spaces not migrated" and I'm once again left with the empty "all.pcap". At this point I stumble onto the workaround that if I delete the empty Pool for "all.pcap", relaunch, then let the migration complete one last time, now I truly do have all my Spaces migrated, including that "all.pcap" one. Now I can relaunch and am no longer prompted about Space migration. It's not shown in the video, but I've confirmed that the `data/spaces/` directory has been deleted at this point, which is the expected end state after Brimcap has successfully completed all the Space migrations.

https://user-images.githubusercontent.com/5934157/118419695-9d352e80-b671-11eb-9f39-36e1ca4edadb.mp4

In conclusion, in a pinch we could document this in the [wiki article](https://github.com/brimdata/brim/wiki/Migration-of-Spaces) and just set expectations that these need to be deleted in order for them to complete. I'm hoping that cancel-in-progress will not be a common user operation, and this Space migration is just a one-time thing, after all. OTOH, if it would be easy for Brimcap to just clean up the empty Pool in the event of a cancel, that might be even better.

**Comment by jameskerr** [comment]
Thanks @phil, yes this matches my experience when developing. I did not discover that workaround you found. If it is helpful for context, the "Some spaces not migrated" notification appears if, after calling the brimcap migrate script, there are still some folders in the data/spaces directory.

**Comment by mattnibs** [comment]
I have built in logic to support the cancellation case. I am curious what the logs from brimcap look like when the cancellation occurs.

**Comment by philrz** [comment]
@mattnibs: I just repeated the test with the Brim scratch build [https://storage.googleapis.com/brimsec-releases/brim/v0.25.0-prerelease-bd5d9930.0/macos/Brim-0.25.0-prerelease-bd5d9930.0.dmg](https://storage.googleapis.com/brimsec-releases/brim/v0.25.0-prerelease-bd5d9930.0/macos/Brim-0.25.0-prerelease-bd5d9930.0.dmg) which uses Brimcap commit `c00eb28` that includes your additional debug info.

```
$ /Applications/Brim.app/Contents/Resources/app.asar.unpacked/zdeps/brimcap -version
Version: v0.0.3-30-gc00eb28
```

I once again canceled when it was in the middle of migrating `all.pcap`. The tl;dr as far as I can tell is that there's no "abort"-specific Brimcap logs that I can see, so maybe it's not being aborted in the way you think it is? In any case, I'm pasting the logs for completeness.

Here's the `main.log`, which includes remnants in the top half from when I was prepping things in the `zqd` of Brim `v0.24.0` before upgrading & migrating.

```
[2021-05-17 19:15:35.052] [info] app paths: getAppPath=/Applications/Brim.app/Contents/Resources/app userData=/Users/phil/Library/Application Support/Brim logs=/Users/phil/Library/Application Support/Brim/logs
[2021-05-17 19:15:35.240] [info] zqd core log /Users/phil/Library/Application Support/Brim/logs/zqd.log
[2021-05-17 19:15:35.240] [info] zqd config /Users/phil/Library/Application Support/Brim/zqd-config.yaml
[2021-05-17 19:15:35.250] [info] spawning zqd: /Applications/Brim.app/Contents/Resources/app/zdeps/zqd listen -l localhost:9867 -data /Users/phil/Library/Application Support/Brim/data/spaces -config /Users/phil/Library/Application Support/Brim/zqd-config.yaml -suricatarunner /Applications/Brim.app/Contents/Resources/app/zdeps/suricata/suricatarunner -suricataupdater /Applications/Brim.app/Contents/Resources/app/zdeps/suricata/suricataupdater -zeekrunner /Applications/Brim.app/Contents/Resources/app/zdeps/zeek/zeekrunner -brimfd=3
[2021-05-17 19:15:42.737] [info] zqd has shutdown
[2021-05-17 19:15:59.841] [info] app paths: getAppPath=/Applications/Brim.app/Contents/Resources/app userData=/Users/phil/Library/Application Support/Brim logs=/Users/phil/Library/Application Support/Brim/logs
[2021-05-17 19:16:00.041] [info] migrations: currentVersion=202101210823 pending=0
[2021-05-17 19:16:00.053] [info] zqd core log /Users/phil/Library/Application Support/Brim/logs/zqd.log
[2021-05-17 19:16:00.054] [info] zqd config /Users/phil/Library/Application Support/Brim/zqd-config.yaml
[2021-05-17 19:16:00.058] [info] spawning zqd: /Applications/Brim.app/Contents/Resources/app/zdeps/zqd listen -l localhost:9867 -data /Users/phil/Library/Application Support/Brim/data/spaces -config /Users/phil/Library/Application Support/Brim/zqd-config.yaml -suricatarunner /Applications/Brim.app/Contents/Resources/app/zdeps/suricata/suricatarunner -suricataupdater /Applications/Brim.app/Contents/Resources/app/zdeps/suricata/suricataupdater -zeekrunner /Applications/Brim.app/Contents/Resources/app/zdeps/zeek/zeekrunner -brimfd=3
[2021-05-17 19:16:05.343] [info] zqd has shutdown
[2021-05-17 19:16:51.489] [info] app paths: getAppPath=/Applications/Brim.app/Contents/Resources/app.asar userData=/Users/phil/Library/Application Support/Brim logs=/Users/phil/Library/Application Support/Brim/logs
[2021-05-17 19:16:51.673] [info] migrations: currentVersion=202101210823 pending=7
[2021-05-17 19:16:51.673] [info] migrations: running
[2021-05-17 19:16:51.676] [info] migrations: currentVersion=202105141312
[2021-05-17 19:16:51.719] [info] zed core log /Users/phil/Library/Application Support/Brim/logs/zlake.log
[2021-05-17 19:16:51.812] [info] spawning zed lake serve: /Applications/Brim.app/Contents/Resources/app.asar.unpacked/zdeps/zed lake serve -l localhost:9867 -R /Users/phil/Library/Application Support/Brim/data/lake -log.level=info -log.filemode=rotate -log.path /Users/phil/Library/Application Support/Brim/logs/zlake.log -brimfd=3
[2021-05-17 19:17:22.302] [error] There was a problem updating the application: HttpError: 404 
"method: GET url: https://github.com/brimdata/brim/releases/download/v0.24.0/latest-mac.yml\n\nPlease double check that your authentication token is correct. Due to security reasons actual status maybe not reported, but 404.\n"
Headers: {
  "server": "GitHub.com",
  "date": "Tue, 18 May 2021 02:17:22 GMT",
  "content-type": "text/plain; charset=utf-8",
  "vary": "X-PJAX, Accept-Encoding, Accept, X-Requested-With",
  "permissions-policy": "interest-cohort=()",
  "cache-control": "no-cache",
  "strict-transport-security": "max-age=31536000; includeSubdomains; preload",
  "x-frame-options": "deny",
  "x-content-type-options": "nosniff",
  "x-xss-protection": "0",
  "referrer-policy": "no-referrer-when-downgrade",
  "expect-ct": "max-age=2592000, report-uri=\"https://api.github.com/_private/browser/errors\"",
  "content-security-policy": "default-src 'none'; base-uri 'self'; connect-src 'self'; form-action 'self'; img-src 'self' data:; script-src 'self'; style-src 'unsafe-inline'",
  "content-encoding": "gzip",
  "content-length": "29",
  "x-github-request-id": "D52B:672E:17456B:2A1BC3:60A323B1"
}
```

Here's the contents of the `zlake.log`:

```
{"level":"info","ts":1621304211.8644228,"msg":"Open files limit raised","limit":24576}
{"level":"info","ts":1621304211.8649569,"msg":"Listening to brim process pipe","fd":"brimfd"}
{"level":"info","ts":1621304211.8695822,"logger":"core","msg":"Started"}
{"level":"info","ts":1621304211.8727071,"logger":"httpd","msg":"Listening","addr":"127.0.0.1:9867"}
{"level":"info","ts":1621304213.624588,"logger":"http.access","msg":"Request completed","request_id":"1sgpiRZt9vmAQ085zN5vOvcsFWc","host":"localhost:9867","method":"GET","proto":"HTTP/1.1","remote_addr":"127.0.0.1:54565","request_content_length":0,"url":"/pool","elapsed":0.000279663,"response_content_length":5,"status_code":200}
{"level":"info","ts":1621304223.1102579,"logger":"http.access","msg":"Request completed","request_id":"1sgpjjHetqGMRqNkBuRdDIZNlzk","host":"localhost:9867","method":"POST","proto":"HTTP/1.1","remote_addr":"127.0.0.1:54570","request_content_length":60,"url":"/pool","elapsed":0.001501652,"response_content_length":58,"status_code":200}
{"level":"info","ts":1621304223.118878,"logger":"http.access","msg":"Request completed","request_id":"1sgpjli6lYf5Y3kgs1MlKeCbxMR","host":"localhost:9867","method":"GET","proto":"HTTP/1.1","remote_addr":"127.0.0.1:54565","request_content_length":0,"url":"/pool","elapsed":0.000015378,"response_content_length":60,"status_code":200}
{"level":"info","ts":1621304225.795242,"logger":"http.access","msg":"Request completed","request_id":"1sgpjsjCgiTVNvYMh7tY8fmnaAl","host":"localhost:9867","method":"POST","proto":"HTTP/1.1","remote_addr":"127.0.0.1:54570","request_content_length":-1,"url":"/pool/1sgpjhDc0ra27YIFq6zYtYZFKj2/log","elapsed":1.215254965,"response_content_length":104,"status_code":200}
{"level":"info","ts":1621304225.804794,"logger":"http.access","msg":"Request completed","request_id":"1sgpjxPhAW0jktRwF16ZYq0vDvU","host":"localhost:9867","method":"POST","proto":"HTTP/1.1","remote_addr":"127.0.0.1:54570","request_content_length":61,"url":"/pool","elapsed":0.00092322,"response_content_length":59,"status_code":200}
{"level":"info","ts":1621304225.8113458,"logger":"http.access","msg":"Request completed","request_id":"1sgpjxcUqYIFu7xp36k0QGelS54","host":"localhost:9867","method":"POST","proto":"HTTP/1.1","remote_addr":"127.0.0.1:54570","request_content_length":-1,"url":"/pool/1sgpk0nbmJFKafhnXFdzSPDhQL7/log","elapsed":0.005084814,"response_content_length":100,"status_code":200}
{"level":"info","ts":1621304225.8139362,"logger":"http.access","msg":"Request completed","request_id":"1sgpjyhshmvMQvP8Nzcuikde8IV","host":"localhost:9867","method":"POST","proto":"HTTP/1.1","remote_addr":"127.0.0.1:54570","request_content_length":57,"url":"/pool","elapsed":0.00115514,"response_content_length":55,"status_code":200}
{"level":"info","ts":1621304225.814641,"logger":"http.access","msg":"Request completed","request_id":"1sgpk0AlzTd7Rcb8P8vaSD47mN7","host":"localhost:9867","method":"GET","proto":"HTTP/1.1","remote_addr":"127.0.0.1:54565","request_content_length":0,"url":"/pool","elapsed":0.000064117,"response_content_length":174,"status_code":200}
{"level":"info","ts":1621304225.825102,"logger":"http.access","msg":"Request completed","request_id":"1sgpjuZ7gMKaTg0ae7YT0mavlZX","host":"localhost:9867","method":"GET","proto":"HTTP/1.1","remote_addr":"127.0.0.1:54565","request_content_length":0,"url":"/pool","elapsed":0.000019911,"response_content_length":174,"status_code":200}
```

And here's the messages on the DevTool console during Space migration:

```
{"msg":"migrating 5 spaces"}
/Applications/Brim.app/Contents/Resources/app.asar/dist/app/legacy/space-migration/space-migrator.js:93 {"space":"wrccdc.pcap","msg":"migration starting"}
/Applications/Brim.app/Contents/Resources/app.asar/dist/app/legacy/space-migration/space-migrator.js:93 {"space":"wrccdc.pcap","msg":"migrating pcap"}
/Applications/Brim.app/Contents/Resources/app.asar/dist/app/legacy/space-migration/space-migrator.js:93 {"space":"wrccdc.pcap","msg":"migrating data"}
/Applications/Brim.app/Contents/Resources/app.asar/dist/app/legacy/space-migration/space-migrator.js:93 {"space":"wrccdc.pcap","msg":"data migration completed"}
/Applications/Brim.app/Contents/Resources/app.asar/dist/app/legacy/space-migration/space-migrator.js:93 {"space":"wrccdc.pcap","msg":"migration successful"}
/Applications/Brim.app/Contents/Resources/app.asar/dist/app/legacy/space-migration/space-migrator.js:93 {"space":"hello.pcapng","msg":"migration starting"}
/Applications/Brim.app/Contents/Resources/app.asar/dist/app/legacy/space-migration/space-migrator.js:93 {"space":"hello.pcapng","msg":"migrating pcap"}
/Applications/Brim.app/Contents/Resources/app.asar/dist/app/legacy/space-migration/space-migrator.js:93 {"space":"hello.pcapng","msg":"migrating data"}
/Applications/Brim.app/Contents/Resources/app.asar/dist/app/legacy/space-migration/space-migrator.js:93 {"space":"hello.pcapng","msg":"data migration completed"}
/Applications/Brim.app/Contents/Resources/app.asar/dist/app/legacy/space-migration/space-migrator.js:93 {"space":"hello.pcapng","msg":"migration successful"}
/Applications/Brim.app/Contents/Resources/app.asar/dist/app/legacy/space-migration/space-migrator.js:93 {"space":"all.pcap","msg":"migration starting"}
/Applications/Brim.app/Contents/Resources/app.asar/dist/app/legacy/space-migration/space-migrator.js:93 {"space":"all.pcap","msg":"migrating pcap"}
```

**Comment by mattnibs** [comment]
Thanks @philrz !

**Comment by philrz** [comment]
Verified in Brim commit `5c6fb2cb` which uses Brimcap commit 5dae4c5 and Zed commit `7adbb36`.

The attached video shows a detailed verification on macOS using scratch artifact [https://storage.googleapis.com/brimsec-releases/brim/v0.25.0-prerelease-5c6fb2cb.0/macos/Brim-0.25.0-prerelease-5c6fb2cb.0.dmg](https://storage.googleapis.com/brimsec-releases/brim/v0.25.0-prerelease-5c6fb2cb.0/macos/Brim-0.25.0-prerelease-5c6fb2cb.0.dmg). It follows the original repro steps, with the post-upgrade migration being canceled in the middle of migrating a Space called `all.pcap`. As shown, once the user selects **Cancel**, after a pause of several seconds while the partially-migrated data is cleaned up, the user is notified and the Pool name disappears from the list. If the user closes and re-opens the app as is done here, the remaining un-migrated Spaces are still seen as present and the user is once again prompted to complete migration (the same pop-up would appear it the user had selected **File > New Window** or **View > Reload**). Once all migrations have completed, on an additional close & relaunch, we confirm that the migration prompt no longer appears, and clicking on the data in the new Pools works as expected.

https://user-images.githubusercontent.com/5934157/119401062-d5e79000-bc8f-11eb-8246-7c5f94fb0dce.mp4

I also confirmed we're covered for all platforms. Here's a verification on Windows using scratch artifact https://storage.googleapis.com/brimsec-releases/brim/v0.25.0-prerelease-5c6fb2cb.0/windows/Brim-Setup-0.25.0-prerelease-5c6fb2cb.0.exe. For variety, here I happen to use the **File > New Window** approach to triggering the remaining migrations after the cancel.

https://user-images.githubusercontent.com/5934157/119403494-4e038500-bc93-11eb-8f3c-7c321c6e5bd2.mp4

Here's verification on Ubuntu 18.04 using scratch artifact https://storage.googleapis.com/brimsec-releases/brim/v0.25.0-prerelease-5c6fb2cb.0/linux/Brim-0.25.0-prerelease-5c6fb2cb.0.deb. For variety, here during the resumed migration I show the convenience of being able to work with the existing Pools while migrations are still occurring.

https://user-images.githubusercontent.com/5934157/119404883-4f35b180-bc95-11eb-9b41-1032216c769b.mp4

Finally, here's verification on CentOS 8 using scratch artifact https://storage.googleapis.com/brimsec-releases/brim/v0.25.0-prerelease-5c6fb2cb.0/linux/Brim-0.25.0-prerelease-5c6fb2cb.0.rpm.

https://user-images.githubusercontent.com/5934157/119406001-fe26bd00-bc96-11eb-8a2b-f43e1469c9e7.mp4

Thanks @mattnibs!
